sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -437,6 +437,14 @@ pi_result _pi_context::initialize() {
   return PI_SUCCESS;
 }
 
+void _pi_context::finalize() {
+  // This function is called when pi_context is deallocated, piContextRelase.
+  // There could be some memory that may have not been deallocated.
+  // For example, zeEventPool could be still alive.
+  if (ZeEventPool && NumEventsLiveInEventPool[ZeEventPool])
```

Comment:
If multiple threads operate on the same context then there is a race condition here. We need to start locking the entire context in PI interfaces operating on a context, similar to what we are already doing for queue (see PiQueueMutex)

Reply:
Done

Reply:
This is better, thanks. I meant that we need to start guarding the entire input context we are operating on, but that's a separate issue, and should be solved separetly.